editor/zls/src/document_symbols.rs

```diff
@@ -0,0 +1,150 @@
+use zydeco_lang::utils::span::{Sp, Span, SpanView};
+use zydeco_lang::zydeco::ZydecoFile;
+
+use crate::text_position::span_to_range;
+
+pub fn handle(
+    document: &FullTextDocument,
+) -> Result<Option<DocumentSymbolResponse>, ResponseError> {
+    let ast = ZydecoFile::parse_src(document.get_content(None), PathBuf::new()).map_err(|err| {
+        ResponseError::new(ErrorCode::REQUEST_FAILED, format!("parsing error: {err}"))
+    })?;
+
+    let symbols = ast
+        .inner
+        .declarations
+        .iter()
+        .flat_map(|decl| declaration_to_symbol(decl, document))
+        .collect();
+    Ok(Some(DocumentSymbolResponse::Nested(symbols)))
+}
+
+fn declaration_to_symbol(
+    decl: &Sp<DeclSymbol<Declaration>>, document: &FullTextDocument,
+) -> Vec<DocumentSymbol> {
+    match &decl.inner.inner {
+        Declaration::Module(module_def) => {
+            let children = module_def
+                .declarations
+                .iter()
+                .flat_map(|decl| declaration_to_symbol(decl, document))
+                .collect();
+
+            match &module_def.name {
+                Some(name_ref) => {
+                    vec![name_to_symbol(
+                        name_ref,
+                        decl.span(),
+                        document,
+                        SymbolKind::MODULE,
+                        children,
+                    )]
+                }
+                None => children,
+            }
+        }
+        Declaration::UseDef(_) => vec![],
+        Declaration::Data(data_def) => {
+            let children =
+                data_def.ctors.iter().map(|ctor| ctorv_to_symbol(&ctor.ctorv, document)).collect();
+
+            vec![name_to_symbol(&data_def.name, decl.span(), document, SymbolKind::ENUM, children)]
+        }
+        Declaration::Codata(codata_def) => {
+            let children = codata_def
+                .dtors
+                .iter()
+                .map(|dtor| dtorv_to_symbol(&dtor.dtorv, document))
+                .collect();
+
+            vec![name_to_symbol(
+                &codata_def.name,
+                decl.span(),
+                document,
+                SymbolKind::CLASS,
+                children,
+            )]
+        }
+        Declaration::Alias(alias_def) => {
+            // TODO: lookup symbol to display correct kind
+            vec![name_to_symbol(&alias_def.name, decl.span(), document, SymbolKind::ENUM, vec![])]
+        }
+        Declaration::Define(def) => {
+            // TODO: handle local definitions
+            let children = vec![];
+
+            let kind = match def.0.params.len() {
+                0 => SymbolKind::CONSTANT,
+                _ => SymbolKind::FUNCTION,
+            };
+
+            vec![name_to_symbol(&def.0.name.0, decl.span(), document, kind, children)]
+        }
+        // TODO: handle main, local definitions
+        Declaration::Main(_) => vec![],
+    }
+}
+
+#[allow(deprecated)]
```

Comment:
Is there anything deprecated?

Reply:
This is an unfortunate side effect of how `lsp_types` uses `#[deprecated]` (see also https://github.com/gluon-lang/lsp-types/issues/226).

The [LSP specified `DocumentSymbol` type](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentSymbol) includes a deprecated field for backwards compatibility. The `lsp_types` (correctly) includes this field in their type, but chose to represent the deprecation in LSP as a deprecation in Rust. As `DocumentSymbol` has no constructor function, we must name the deprecated field to initialize the struct, which causes a deprecation warning.

There are a few other warts in `lsp_types`, but I'm not aware of a competing crate at present.

Reply:
I see... Maybe we can wrap around `DocumentSymbol` by presenting a constructor function and only use `#[allow(deprecated)]` there. Not a major issue though. Feel free to resolve the conversation.

Reply:
Just pushed a commit with a constructor function to minimize `#[allow(deprecated)]`.